src/ocean/util/container/btree/BTreeMap.d

```diff
@@ -0,0 +1,664 @@
+
+    public ElementType findMinimum (out bool found)
+    {
+        return this.impl.findMinimum(found);
+    }
+
+    /******************************************************************************
+
+        Searches the tree for the element and returns the pointer to it, or null
+        if not found.
+
+        Params:
+            needle = element to find in a tree.
+            found_element = out parameter to be filled with the copy of the found
+                element.
+
+        Returns:
+            true if the element is found, false otherwise
+
+    *******************************************************************************/
+
+    public ElementType search (KeyType key, out bool found_element)
+    {
+        return this.impl.search(key, found_element);
+    }
+
+    /***********************************************************************
+
+        Recursive inorder iteration. Note that, in case the tree is changed
+        during the iteration, iteration will halt.
+
+        Params:
+            dg = opApply's delegate
+
+        Returns:
+            return value of dg
+
+    ***********************************************************************/
+
+    public int opApply (int delegate (ref KeyType value, ref ElementType) dg)
+    {
+        return this.impl.inorder(dg);
+    }
+
+    /***********************************************************************
+
+        Recursive inorder iteration. Note that, in case the tree is changed
+        during the iteration, iteration will halt.
+
+        Params:
+            dg = opApply's delegate
+
+        Returns:
+            return value of dg
+
+    ***********************************************************************/
+
+    public int opApply (int delegate (ref ElementType) dg)
+    {
+        return this.impl.inorder(dg);
+    }
+
+
+    /// Convenience alias for the implementation
+    package alias BTreeMapImplementation!(KeyType, ElementType, tree_degree)
+        Implementation;
+
+    /// Convenience alias for the key type
+    package alias TreeKeyType KeyType;
+
+    /// Convenience alias for the element type
+    package alias TreeElementType ElementType;
+
+    /// Private implementation
+    package Implementation impl;
+}
+
+/*******************************************************************************
+
+    Constructor-like method. Constructs the BTreeMap and initializes it.
+
+    Params:
+        KeyType = type of the key.
+        ElementType = type of the element to store.
+        tree_degree = degree of the tree (refer to the documentation)
+        memManager = memory management strategy to use
+
+******************************************************************************/
+
+public BTreeMap!(KeyType, ElementType, tree_degree) createBTreeMap
+    (KeyType, ElementType, int tree_degree)(IMemManager allocator = mallocMemManager)
+{
+    BTreeMap!(KeyType, ElementType, tree_degree) tree;
+    tree.init(allocator);
+    return tree;
+}
+
+
+///
```

Comment:
You can `@disable this` for actual struct and prepare it in constructor function from `T.init` state.